rfc/20241003-temporary-files.md

```diff
@@ -0,0 +1,447 @@
+
+    If an attacker manages to modify the upstream copy of a package and also publishes a new signature covering their modified copy using a different private key, the checksums recorded in the dependency lock file allow OpenTofu to detect and report the inconsistency, so the operator can investigate further and decide how to proceed.
+
+- If running plan and apply separately using a saved plan file, OpenTofu verifies that the plan is being applied with the same providers that it was created with. This is important for correctness, because the `ApplyResourceChange` RPC for a particular provider plugin is only required to accept as input a verbatim copy of something previously returned by `PlanResourceChange`, and providers can potentially misbehave in unpredictable ways if that guarantee isn't maintained.
+
+At the time of originally designing the dependency lock file, I had intended it to cover both provider _and_ module packages. However, while external provider packages had been designed from the outset to be treated as immutable (due to them being cryptographically signed), the module installation mechanism has been around for considerably longer and was not designed with such ideas in mind.
+
+By the time I came to designing and implementing the dependency lock file, the prevailing patterns for existing modules already included these workarounds of using `path.module` or `path.cwd` to directly modify a module's own directory, and so attempts to capture and record checksums for module packages were very troublesome and often generated spurious inconsistency errors in practice due to the general assumption that modules were free to modify themselves arbitrarily at runtime (along with other reasons that I'll touch on below for context).
+
+Although the introduction of `path.temp` cannot immediately compell all existing modules to stop self-modifying, it does at least establish an alternative pattern that addresses what is by far the most common reason for a module to modify itself today, hopefully setting us on a path where in future a greater percentage of the most commonly-used modules are distributed in packages that can be assumed to be immutable, and thus be reliably covered by checksums in the dependency lock file.
+
+However, I do want to be clear that this issue is far from the _only_ change that would be required for OpenTofu to provide the same full set of guarantees for module packages as it does for provider packages (none of which are related to `path.temp`, but included here for context):
+
+- The provider registry protocol does not have any mechanism for authors to sign their packages, so although we can generate checksums to determine whether a module package is still the same as the last time we saw it, we would not have any comparable idea of what is the "official" content of a module package.
```

Comment:
This item should start with "The _module_ registry protocol". (This point is trying to point out a significant difference in the trust/security model of module registries vs. provider registries.)